graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/tracing/FederatedTracingInstrumentation.java

```diff
@@ -0,0 +1,228 @@
+     */
+    private static class FederatedTracingState implements InstrumentationState {
+        private final Instant startRequestTime;
+        private final long startRequestNanos;
+        private final LinkedHashMap<ExecutionPath, Reports.Trace.Node.Builder> nodesByPath;
+
+        private FederatedTracingState() {
+            // record start time when creating instrumentation state for a request
+            startRequestTime = Instant.now();
+            startRequestNanos = System.nanoTime();
+
+            nodesByPath = new LinkedHashMap<>();
+            nodesByPath.put(ExecutionPath.rootPath(), Reports.Trace.Node.newBuilder());
+        }
+
+        Reports.Trace toProto() {
+            return Reports.Trace.newBuilder()
+                    .setStartTime(getStartTimestamp())
+                    .setEndTime(getNowTimestamp())
+                    .setDurationNs(getDuration())
+                    .setRoot(nodesByPath.get(ExecutionPath.rootPath()))
+                    .build();
+        }
+
+        /**
+         * Adds node to nodesByPath and recursively ensures that all parent nodes exist.
+         */
+        void addNode(ExecutionStepInfo stepInfo, long startFieldNanos, long endFieldNanos, List<GraphQLError> errors) {
+            ExecutionPath path = stepInfo.getPath();
+            Reports.Trace.Node.Builder parent = getParent(path);
+
+            Reports.Trace.Node.Builder node = parent.addChildBuilder()
+                    .setStartTime(startFieldNanos)
+                    .setEndTime(endFieldNanos)
+                    .setParentType(simplePrint(stepInfo.getParent().getUnwrappedNonNullType()))
+                    .setType(stepInfo.simplePrint())
+                    .setResponseName(stepInfo.getResultKey());
+
+            String originalFieldName = stepInfo.getField().getName();
+
+            // set originalFieldName only when a field alias was used
+            if (!originalFieldName.equals(stepInfo.getResultKey())) {
+                node.setOriginalFieldName(originalFieldName);
+            }
+
+            errors.forEach(error -> {
+                Reports.Trace.Error.Builder builder = node.addErrorBuilder().setMessage(error.getMessage());
+                error.getLocations().forEach(location -> builder.addLocationBuilder().setColumn(location.getColumn()).setLine(location.getLine()));
+            });
+
+            nodesByPath.put(path, node);
+        }
+
+        Reports.Trace.Node.Builder getParent(ExecutionPath path) {
+            List<Object> pathParts = path.toList();
+            ExecutionPath parentPath = ExecutionPath.fromList(pathParts.subList(0, pathParts.size() - 1));
+
+            if (!nodesByPath.containsKey(parentPath)) {
+                Reports.Trace.Node.Builder missingParent = getParent(parentPath).addChildBuilder();
+
+                // Missing parents are always list items, so we need to add the `index` field to them
```

Comment:
Why is this true? Is this relying on the way that GraphQL-Java executes? And why isn't it true for list items?

Reply:
I assume it's based on the assumption that execution goes from top to bottom.

Note though that GraphQL types can be nested lists – is that what you're asking about in the second sentence?

Reply:
that makes sense, and I think that's true, but I'm questioning whether that's a safe assumption to bake in and if not, how we might avoid it

Reply:
There is an instrumentation hook called [beginFieldListComplete](https://github.com/graphql-java/graphql-java/blob/b5781e3bbaeababe4bb916cf6acba6763fd7c875/src/main/java/graphql/execution/instrumentation/Instrumentation.java#L154). It seemed promising because there's a `getCurrentListIndex()` as part of the arguments to the hook. But it doesn't work: it's called once per list, not once per list item. ([Calling code here](https://github.com/graphql-java/graphql-java/blob/735ce0626e608ad1c6530924f8076851aa18f70b/src/main/java/graphql/execution/ExecutionStrategy.java#L500)).

I think this is the best bet. I didn't consider nested lists, so I added a test for them and it seems to work fine!

If y'all can figure out another way to create the indexed nodes, I'm all ears. Maybe I'll post something on the graphql-java spectrum to see if one of the maintainers have any ideas.

Reply:
https://spectrum.chat/graphql-java/general/instrumentation-hook-for-list-items~534fa292-73b0-4cce-8832-5bbda66dcfb3